File: Fortro-Engine/src/api/verifier.rs
use axum::{
    extract::{Json, Path, State, Query},
    http::StatusCode,
    routing::{get, post, put},
    Router,
};
use serde::{Deserialize};
use serde_json::json;
use std::collections::HashMap;

use crate::error::AppError;
use crate::models::{PresentationStatus, CredentialRequirement, AccessLevel, ExpirationPolicy};
use crate::services::AppState;
use crate::services::verifier::{CreatePresentationRequestRequest, VerifyPresentationRequest};

/// Create verifier routes
pub fn routes() -> Router<AppState> {
    Router::new()
        // Presentation requests
        .route("/requests", post(create_presentation_request))
        .route("/requests/:id", get(get_presentation_request))
        
        // Presentations
        .route("/presentations", get(list_presentations))
        .route("/presentations/:id", get(get_presentation))
        .route("/presentations/:id/verify", post(verify_presentation))
        .route("/presentations/:id/status", put(update_presentation_status))
        
        // Consent management
        .route("/consents", get(list_consents))
        .route("/consents/request", post(request_consent))
        .route("/consents/check", post(check_consent))
        
        // QR code generation
        .route("/qr/presentation-request", post(generate_presentation_request_qr))
        
        // Statistics
        .route("/:did/statistics", get(get_verifier_statistics))
}

/// Create presentation request handler
async fn create_presentation_request(
    State(state): State<AppState>,
    Json(request): Json<CreatePresentationRequestRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let response = verifier_service.create_presentation_request(request).await?;

    Ok((
        StatusCode::CREATED,
        Json(json!({
            "success": true,
            "message": "Presentation request created successfully",
            "request": response.request,
            "qr_code_data": response.qr_code_data,
        })),
    ))
}

/// Get presentation request handler
async fn get_presentation_request(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let presentation_service = state.presentation_service();
    let request = presentation_service.get_presentation_by_id(&id).await?
        .ok_or_else(|| AppError::NotFoundError(format!("Presentation request with ID {} not found", id)))?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "request": request,
        })),
    ))
}

/// List presentations handler
async fn list_presentations(
    State(state): State<AppState>,
    Query(params): Query<HashMap<String, String>>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_did = params.get("verifier_did")
        .ok_or_else(|| AppError::ValidationError("verifier_did parameter is required".to_string()))?;
    
    let verifier_service = state.verifier_service();
    let presentations = verifier_service.get_presentations_by_verifier(verifier_did).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "presentations": presentations,
        })),
    ))
}

/// Get presentation handler
async fn get_presentation(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let presentation = verifier_service.get_presentation_by_id(&id).await?
        .ok_or_else(|| AppError::NotFoundError(format!("Presentation with ID {} not found", id)))?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "presentation": presentation,
        })),
    ))
}

/// Verify presentation handler
async fn verify_presentation(
    State(state): State<AppState>,
    Json(request): Json<VerifyPresentationRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let result = verifier_service.verify_presentation(request).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "verification_result": result,
        })),
    ))
}

/// Update presentation status request
#[derive(Debug, Deserialize)]
pub struct UpdatePresentationStatusRequest {
    pub verifier_did: String,
    pub status: PresentationStatus,
}

/// Update presentation status handler
async fn update_presentation_status(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(request): Json<UpdatePresentationStatusRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let success = verifier_service.update_presentation_status(&id, &request.verifier_did, request.status).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": success,
            "message": "Presentation status updated successfully",
        })),
    ))
}

/// List consents handler
async fn list_consents(
    State(state): State<AppState>,
    Query(params): Query<HashMap<String, String>>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_did = params.get("verifier_did")
        .ok_or_else(|| AppError::ValidationError("verifier_did parameter is required".to_string()))?;
    
    let verifier_service = state.verifier_service();
    let consents = verifier_service.get_consents_for_verifier(verifier_did).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "consents": consents,
        })),
    ))
}

/// Request consent request
#[derive(Debug, Deserialize)]
pub struct RequestConsentRequest {
    pub verifier_did: String,
    pub user_did: String,
    pub purpose: String,
    pub data_categories: Vec<String>,
    pub access_level: AccessLevel,
    pub expiration_policy: ExpirationPolicy,
    pub expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

/// Request consent handler
async fn request_consent(
    State(state): State<AppState>,
    Json(request): Json<RequestConsentRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let consent = verifier_service.request_consent(
        &request.verifier_did,
        &request.user_did,
        &request.purpose,
        request.data_categories,
        request.access_level,
        request.expiration_policy,
        request.expires_at,
    ).await?;

    Ok((
        StatusCode::CREATED,
        Json(json!({
            "success": true,
            "message": "Consent request created successfully",
            "consent": consent,
        })),
    ))
}

/// Check consent request
#[derive(Debug, Deserialize)]
pub struct CheckConsentRequest {
    pub verifier_did: String,
    pub user_did: String,
    pub purpose: String,
}

/// Check consent handler
async fn check_consent(
    State(state): State<AppState>,
    Json(request): Json<CheckConsentRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let has_consent = verifier_service.check_consent(
        &request.verifier_did,
        &request.user_did,
        &request.purpose,
    ).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "has_consent": has_consent,
        })),
    ))
}

/// Generate presentation request QR code request
#[derive(Debug, Deserialize)]
pub struct GeneratePresentationRequestQrRequest {
    pub verifier_did: String,
    pub required_credentials: Vec<CredentialRequirement>,
    pub presentation_type: String,
    pub purpose: String,
    pub callback_url: Option<String>,
    pub expires_at: Option<chrono::DateTime<chrono::Utc>>,
}

/// Generate presentation request QR code handler
async fn generate_presentation_request_qr(
    State(state): State<AppState>,
    Json(request): Json<GeneratePresentationRequestQrRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let qr_data = verifier_service.generate_presentation_request_qr(
        &request.verifier_did,
        request.required_credentials,
        &request.presentation_type,
        &request.purpose,
        request.callback_url,
        request.expires_at,
    ).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "qr_data": qr_data,
        })),
    ))
}

/// Get verifier statistics handler
async fn get_verifier_statistics(
    State(state): State<AppState>,
    Path(did): Path<String>,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let verifier_service = state.verifier_service();
    let statistics = verifier_service.get_verifier_statistics(&did).await?;

    Ok((
        StatusCode::OK,
        Json(json!({
            "success": true,
            "statistics": statistics,
        })),
    ))
}

File: sphyre-holder/src/lib/api/services/verifier.ts
import { api } from '../client';
import { API_ENDPOINTS } from '../config';

export interface PresentationRequest {
  id: string;
  verifierDid: string;
  name: string;
  description?: string;
  requestedAttributes: {
    [key: string]: {
      name: string;
      required: boolean;
      condition?: {
        type: 'equals' | 'greaterThan' | 'lessThan' | 'includes';
        value: unknown;
      };
    };
  };
  expiresAt?: string;
  status: 'active' | 'expired' | 'revoked';
  createdAt: string;
  updatedAt: string;
}

export interface Presentation {
  id: string;
  requestId: string;
  holderDid: string;
  status: 'pending' | 'verified' | 'rejected' | 'expired';
  verifiedAt?: string;
  rejectedReason?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Consent {
  id: string;
  verifierDid: string;
  holderDid: string;
  scope: string[];
  expiresAt?: string;
  status: 'active' | 'revoked' | 'expired';
  createdAt: string;
  updatedAt: string;
}

export const verifierService = {
  // Presentation Requests
  async createPresentationRequest(
    data: Omit<PresentationRequest, 'id' | 'status' | 'createdAt' | 'updatedAt'>,
    token?: string
  ): Promise<PresentationRequest> {
    return api.post<PresentationRequest>(API_ENDPOINTS.VERIFIERS.REQUESTS, data, { token });
  },

  async getPresentationRequest(requestId: string, token?: string): Promise<PresentationRequest> {
    return api.get<PresentationRequest>(`${API_ENDPOINTS.VERIFIERS.REQUESTS}/${requestId}`, { token });
  },

  // Presentations
  async listPresentations(params?: {
    requestId?: string;
    holderDid?: string;
    status?: string;
    page?: number;
    limit?: number;
  }, token?: string): Promise<{ items: Presentation[]; total: number }> {
    return api.get<{ items: Presentation[]; total: number }>(
      API_ENDPOINTS.VERIFIERS.PRESENTATIONS, 
      { params, token }
    );
  },

  async getPresentation(presentationId: string, token?: string): Promise<Presentation> {
    return api.get<Presentation>(`${API_ENDPOINTS.VERIFIERS.PRESENTATIONS}/${presentationId}`, { token });
  },

  async verifyPresentation(presentationId: string, token?: string): Promise<{ valid: boolean; message?: string }> {
    return api.post<{ valid: boolean; message?: string }>(
      `${API_ENDPOINTS.VERIFIERS.PRESENTATIONS}/${presentationId}/verify`,
      {},
      { token }
    );
  },

  async updatePresentationStatus(
    presentationId: string, 
    status: 'verified' | 'rejected', 
    reason?: string,
    token?: string
  ): Promise<Presentation> {
    return api.put<Presentation>(
      `${API_ENDPOINTS.VERIFIERS.PRESENTATIONS}/${presentationId}/status`,
      { status, reason },
      { token }
    );
  },

  // Consent Management
  async listConsents(params?: {
    verifierDid?: string;
    holderDid?: string;
    status?: string;
    page?: number;
    limit?: number;
  }, token?: string): Promise<{ items: Consent[]; total: number }> {
    return api.get<{ items: Consent[]; total: number }>(
      API_ENDPOINTS.VERIFIERS.CONSENTS,
      { params, token }
    );
  },

  async requestConsent(
    data: {
      verifierDid: string;
      holderDid: string;
      scope: string[];
      purpose: string;
      expiresInDays?: number;
    },
    token?: string
  ): Promise<{ consent: Consent; requiresUserApproval: boolean }> {
    return api.post<{ consent: Consent; requiresUserApproval: boolean }>(
      API_ENDPOINTS.VERIFIERS.REQUEST_CONSENT,
      data,
      { token }
    );
  },

  async checkConsent(
    data: {
      verifierDid: string;
      holderDid: string;
      scope: string[];
    },
    token?: string
  ): Promise<{ hasConsent: boolean; consent?: Consent }> {
    return api.post<{ hasConsent: boolean; consent?: Consent }>(
      API_ENDPOINTS.VERIFIERS.CHECK_CONSENT,
      data,
      { token }
    );
  },

  // QR Code Generation
  async generatePresentationRequestQr(
    data: {
      requestId: string;
      callbackUrl?: string;
      expiresInHours?: number;
    },
    token?: string
  ): Promise<{ qrCode: string; deepLink: string; expiresAt: string }> {
    return api.post<{ qrCode: string; deepLink: string; expiresAt: string }>(
      API_ENDPOINTS.VERIFIERS.QR,
      data,
      { token }
    );
  },

  // Statistics
  async getVerifierStatistics(did: string, token?: string) {
    return api.get(API_ENDPOINTS.VERIFIERS.STATISTICS(did), { token });
  },
};

export default verifierService;
